[WIP] Implement dynamic contract calls

Robbepop opened the pull request:
Implements https://github.com/paritytech/ink/issues/631.

hugopeixoto commented:
Does the scope of this PR include the "As Storage Entity" functionality from #631?
I don't know much about ink internals so it's a bit hard for me to map it to the concepts in the todo list.

Robbepop commented:
> Does the scope of this PR include the "As Storage Entity" functionality from #631?
> I don't know much about ink internals so it's a bit hard for me to map it to the concepts in the todo list.

Is this a very much needed feature? Actually I would like to try to keep the scope of the initial MVP small.

hugopeixoto commented:
> Is this a very much needed feature? Actually I would like to try to keep the scope of the initial MVP small.

That's alright, I was sincerely asking if it's part of it or not, not requesting it to be included.

Robbepop commented:
> > Is this a very much needed feature? Actually I would like to try to keep the scope of the initial MVP small.
> 
> That's alright, I was sincerely asking if it's part of it or not, not requesting it to be included.

The part should open up possibility to add this feature at a later point in time. However, there are still some open questions as to how the storage-related parts should syntactically look like. There are some ways, non of them obviously the best.

xgreenx commented:
Do you have ETA for this cool change?(=

Robbepop commented:
> Do you have ETA for this cool change?(=

The latest commit no longer produces compile errors in the new codegen. So the new codegen basically works now. However, plenty of stuff still needs to be implemented. This whole thing is much more than just the new trait system as it will feature quite a bit other stuff, too. Will write about this more soon.

xgreenx commented:
Could you describe a little bit how static dependency will work, please?

Robbepop commented:
> Could you describe a little bit how static dependency will work, please?

Best described by this delegator example:
https://gist.github.com/Robbepop/eb32d8b7306681243461520c24b61d0f

Robbepop commented:
I just updated to `trybuild 1.0.49` which made it necessary to adjuts lots of UI tests.
Due to the fact that I had to touch all UI tests anyways I also moved them to where they actually belong.
Sorry about the huge diff, @ascjones . However, it is really just moves and trivial adjustments so just ignore mostly.

Robbepop commented:
> @Robbepop There are a number of changes to the way how ink! works in this PR and it would be helpful for the review to have the "Unreleased" section in `RELEASES.md` updated as well. Since we plan on a short-term release candidate as well, we need to do this soon anyway. Could you add this to the PR?

Done. However, I just saw that some of the things that have been implemented as side-PRs have not been included properly into the `RELEASES.md` for the ink! 3.0-rc6. E.g. addition of the new proc. macros `blake2`, `selector_id` and `selector_bytes`.
